### DISCUSSION TITLE
How to fork the mainnet and test with existing deployed contracts

### DISCUSSION THREAD

**Discussion by djm07073** [zkSync-Community-Hub/zksync-developers#112] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I want to test my contract, interacting with deployed Contract.(ex.iZumiswap).

So I follow "https://era.zksync.io/docs/tools/testing/era-test-node.html#configuring-tests" to test in local network.

But I don't know...Can you help me?

### What is the specific issue or error you're encountering?

```
import { expect } from "chai";
import { Wallet, Provider, Contract } from "zksync-web3";
import * as hre from "hardhat";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
import { CONFIGS } from "../config/address";
import {
  IWETH9,
  IZiMenu,
  IiZiSwapFactory,
  IiZiSwapPool,
} from "../typechain-types";
import { FACTORY_ABI, LMANAGER_ABI, POOL_ABI, WETH_ABI } from "../abi/abi";
import { ethers } from "ethers";

const PK = "0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110";

async function deployMenu(deployer: Deployer) {
  const artifact = await deployer.loadArtifact("IZiMenu");
  return await deployer.deploy(artifact, [
    CONFIGS.zksync.izumiSwap.meta.liquidityManager,
    CONFIGS.zksync.izumiSwap.factory,
  ]);
}
async function getFactory(addresss: string) {
  return new Contract(addresss, FACTORY_ABI);
}
async function getPool(addresss: string) {
  return new Contract(addresss, POOL_ABI);
}
async function getLmanager(addresss: string) {
  return new Contract(addresss, LMANAGER_ABI);
}
async function getWETH(addresss: string) {
  return new Contract(addresss, WETH_ABI);
}

describe("IZiMenu", () => {
  let deployer: Deployer;
  let wallet: Wallet;
  let provider: Provider;
  let menu: IZiMenu;
  let factory: IiZiSwapFactory;
  let pool: IiZiSwapPool;
  let weth: IWETH9;
  before("Setting", async () => {
    provider = Provider.getDefaultProvider();
    wallet = new Wallet(PK, provider);
    deployer = new Deployer(hre, wallet);
    factory = (await getFactory(
      CONFIGS.zksync.izumiSwap.factory
    )) as IiZiSwapFactory;
    weth = (await getWETH(CONFIGS.zksync.izumiSwap.wETH)) as IWETH9;
    await weth.connect(wallet).deposit({ value: ethers.utils.parseEther("1") });
    pool = (await getPool(CONFIGS.zksync.pool.USDC_WETH)) as IiZiSwapPool;
    //const state = await pool.state();
    //console.log(state);
  });

  it("deploy menu", async () => {
    menu = (await deployMenu(deployer)) as IZiMenu;
  });
});
```

```
export const CONFIGS = {
  zksync: {
    chainId: 324,
    blockTime: 2,
    izumiSwap: {
      chainId: 324,
      projectName: "IZumi Swap",
      feeTiers: [100n, 400n, 2000n, 10000n],
      factory: "0x575Bfc57B0D3eA0d31b132D622643e71735A6957",
      wETH: "0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6",
      meta: {
        swap: "0x943ac2310D9BC703d6AB5e5e76876e212100f894",
        liquidityManager: "0x483FDE31bcE3DCc168E23a870831b50Ce2cCd1F1",
        toaster: ethers.constants.AddressZero, // TODO: add toaster address
        menu: ethers.constants.AddressZero,
      },
    },
    tokens: {
      USDC: "0x3355df6D4c9C3035724Fd0e3914dE96A5a83aaf4",
      WETH: "0x5AEa5775959fBC2557Cc8789bC1bf90A239D9a91",
    },
    pool: {
      USDC_WETH: "0x43ff8a10b6678462265b00286796e88f03c8839a",
    },
  },
```
```
import { HardhatUserConfig } from "hardhat/config";
import "@nomiclabs/hardhat-ethers";
import "@typechain/hardhat";
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
import "@matterlabs/hardhat-zksync-verify";

// dynamically changes endpoints for local tests
const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://localhost:8011",
        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://testnet.era.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {
      optimizer: {
        enabled: true,
        mode: "3",
      },
    },
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      chainId: 280,
      gasMultiplier: 0,
      initialBaseFeePerGas: 0,
    },
    zkSyncTestnet,
  },
  solidity: {
    version: "0.8.17",
  },
};

export default config;
```

The code above is part of the test code. 

First, I turned on localnetwork through `era_test_node fork mainnet` and modified `hardhat.config.ts` and `script` as shown in the documentation. Then, I ran `yarn test`, but unlike what I thought, I got an error message `TypeError: Cannot read properties of null (reading 'call')` in the `pool.state()` code, so I commented out the existing code to check if `WETH` works well, and added `await weth. connect(wallet).deposit({ value: ethers.utils.parseEther("1") });` and tried to run it, but got the error `Error: missing response (requestBody="{\"method\":\"eth_gasPrice\",\"params\":[],\"id\":42,\"jsonrpc\":\"2. 0\"}", requestMethod="POST", serverError={"errno":-61,"code":"ECONNREFUSED","syscall":"connect","address":"127.0.0.1","port":3050}, url="http://localhost:3050", code=SERVER_ERROR, version=web/5.7.1)` message came up. 

I need help...:(

### Can you share the error messages or logs you're receiving, if any?

.

### Have you made any recent changes to the contract before encountering this issue?

.

### Are there any external libraries or contracts that your contract interacts with?

.

### Can you provide the relevant portions of your contract code where the issue is occurring?

.

### Have you tried to isolate the problem, and if so, what were the results?

.

### What steps have you already taken to try to resolve the issue?

.

**Comment by uF4No** [comment, marked as answer]
Looks like your provider is pointing to `http://localhost:3050` instead of `http://localhost:8011` (where the the in-memory node is running). Try replacing this:

```
  before("Setting", async () => {
    provider = Provider.getDefaultProvider();
```

with:

```
  before("Setting", async () => {
    provider = new Provider('http://localhost:8011');
```

Also, please check if you have the latest version of hardhat-zksync-deploy (0.6.4) as it includes a fix related with the providers it uses behind the scenes.

Let us know if that works or if you get a different output.

**Reply by djm07073** [reply]
thank you very much! `await weth.connect(wallet).deposit({ value: ethers.utils.parseEther("1") });` works well!

**Comment by threadaqged** [comment]
It seems like you're encountering several issues while trying to test your contract on a local zkSync network using Hardhat. The errors you're facing could be related to various factors such as incorrect configuration, network connectivity, or issues with the zkSync test node setup.

To troubleshoot:

Double-check your local network setup to ensure it matches the configurations specified in the zkSync documentation.
Verify that your Hardhat configuration (hardhat.config.ts) is correctly set up to connect to the zkSync testnet.
Ensure that your zkSync test node is running and accessible locally.
Review the error messages carefully to identify specific issues such as connection errors or null values being returned.
Consider reaching out to the zkSync community or support channels for assistance, as they may provide additional insights or troubleshooting tips.
By systematically addressing these potential issues, you should be able to resolve the errors and successfully test your contract on the local zkSync network.

**Comment by adambold** [comment]
It seems like there are several issues you're encountering while testing your contract with zkSync. Here are some steps you can take to troubleshoot:

Ensure that your local node setup is correct and running smoothly. Double-check the configurations and settings you applied according to the zkSync documentation.

Check if the contracts you're interacting with (IZiMenu, IiZiSwapFactory, IiZiSwapPool, IWETH9) have been deployed correctly and are accessible on the local network.

Verify that your test environment is properly configured in your hardhat.config.ts file, especially the zkSyncTestnet network configuration.

Address the specific error messages you're encountering. For example, the "TypeError: Cannot read properties of null (reading 'call')" may indicate that the pool.state() call is not functioning as expected. Review the logic in that part of your code and ensure it's interacting correctly with the deployed contracts.

Investigate the "missing response" error related to eth_gasPrice. It could be caused by misconfigurations or connectivity issues with the local node. Ensure that the node is running and accessible on the specified port.

Consider adding more logging and debugging statements throughout your code to pinpoint where exactly the issues are occurring.

By following these steps and systematically addressing each error or issue, you should be able to make progress in debugging your contract testing setup with zkSync.

**Comment by tathimustingjason** [comment]
To test your contract locally with zksync, ensure that your local test node is properly configured and running. Check for any errors in your configuration or setup, such as incorrect URLs or missing dependencies. Additionally, ensure that your contract interactions are correctly formatted and that you have proper error handling in place to debug any issues that arise.

**Comment by seimabirthcan** [comment]
It seems like there are multiple issues in your test setup. 
 Firstly, the `pool.state()` call is resulting in a TypeError, indicating a problem with the `pool` object.
 Secondly, the error message `Error: missing response` suggests a connection issue with the local node.
 Try debugging each issue separately: 
  For the TypeError, ensure that the `pool` object is properly initialized and that the contract ABI is correctly defined.
  For the connection issue, verify that the local node is running and accessible at the specified endpoint (localhost:3050).
 Check your configurations and contract interactions to address these issues.
